![LU Logo](https://www.df.lu.lv/fileadmin/user_upload/LU.LV/Apaksvietnes/Fakultates/www.df.lu.lv/Par_mums/Logo/DF_logo/01_DF_logo_LV.png)

# Week 4 - korteži, funkcijas, vārdnīcas, kopas






## Kopsavilkums

Mēs apskatīsim sekojošos tematus:

* sarakstu izveide ar īso sintaksi - (list comprehension angliski)
* korteži - nemaināmi saraksti (tuples angliski)
* funkcijas - koda strukturēšanai un atkārtotas izmantošanas nolūkiem
* vārdnīcas - atslēgu-vērtību pāri (dictionary angliski)
* kopas - nesakārtotas unikālu elementu kolekcijas


## Nepieciešamās zināšanas

* Python sintakses pamati
* Python pamata datu tipi - strings, integers, floats, booleans
* Python pamata operatori - +, -, *, /, %, //, **
* zarošanās konstrukcijas - if, elif, else
* cikli: for un while

## Nodarbības mērķi

Nodarbības beigās studenti varēs:

* Izveidot sarakstus ar īso sintaksi
* Izveidot un lietot kortežus
* Izveidot un lietot funkcijas, ar un bez argumentiem
* Izveidot un lietot vārdnīcas
* Izveidot un lietot kopas

In [1]:
# import komandas parasti tiek liktas programmas sākumā

# Python versijas pārbaude
import sys
print(f"Python version: {sys.version}")

Python version: 3.12.5 (tags/v3.12.5:ff3bc82, Aug  6 2024, 20:45:27) [MSC v.1940 64 bit (AMD64)]


## 1. tēma - Sarakstu izveidošana ar īso sintaksi (list comprehension)

In [2]:
# sarakstu izveide ir bieži sastopama darbība
# līdz ar to var būt ērti izmantot īpašu sintaksi

# apskatīsim sarakstu izveidi ar standarta ciklu palīdzību
squares = []

for x in range(10):
    squares.append(x**2)

print(f"squares: {squares}")

squares: [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [2]:
# mēs varam to pašu izdarīt ar īso sintaksi - list comprehension 
also_squares = [x**2 for x in range(10)]

print(f"also_squares: {also_squares}")

also_squares: [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [3]:
# filtrēšana izmantojot īso sintaksi
# mēs varam pievienot if nosacījumu, lai filtrētu sarakstu, to veidojot
# piemēram mēs varam izveidot sarakstu tikai ar pāra skaitļu kvadrātiem
even_squares = [x**2 for x in range(10) if x % 2 == 0]

print(even_squares)

[0, 4, 16, 36, 64]


### Sarakstu izveidošanas ar īso sintaksi pilnais apraksts

```python
[expression for item in list if conditional]
```

Docs: https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions


### Īsās sintakses salīdzinājums ar parastu for ciklu

Kad lietot īso sintaksi:
* Kad vēlamies izveidojot jaunu sarakstu, izmantojot esošu sarakstu
* Kad gribam pielietot funkciju visiem saraksta elementiem
* Kad gribam filtrēt saraksta elementus

Kad lietot parasto for ciklu:
* Kad vēlamies veikt sarežģītākas darbības ar katru saraksta elementu


## 2. tēma - Korteži

### Korteži - nemaināmi saraksti

- Definējam izmantojot iekavas, atdalam ar komatu.
  - `new_tuple = (1, 2, "string")` 
- Korteži ir  *nemaināmi* - *immutable* , pēc kortežu definēšanas to elementus nevar mainīt.
- Korteži ir noderīgi grupējot elementus, kuru saturu nevēlamies mainīt piemēram konkrētuIP adresi. 
- Korteži arī tiek izmantoti lai atgriestu vairākas vērtības no funkcijas (par funkcijām runāsim sīkāk šajā lekcijā).

Tehniska piezīme: ja kāds korteža elements ir maināms objekts (piemēram saraksts), tad šī objekta saturu var mainīt ar attiecīgām metodēm. Tomēr pats objekts nevar tikt mainīts uz citu objektu.


In [4]:
# izveidosim kortežu ar IP adresi un portu
ip_port = ('192.168.0.1', 80)

# cik garš ir kortežs
print("Tuple length", len(ip_port))

# pirmā korteža elementa vērtība
print("First element of the tuple", ip_port[0])

# otra korteža elementa vērtība
print("Second element of the tuple", ip_port[1])



Tuple length 2
First element of the tuple 192.168.0.1
Second element of the tuple 80


In [5]:
# vai skaitlis 80 ir mūsu kortežā?
print("Is 80 in the tuple?", 80 in ip_port)

# vai skaitlis 192 ir mūsu kortežā?
print("Is 192 in the tuple?", 192 in ip_port) 
# kāpēc šis atgriež False?

# vai teksts "192" ir pirmajā korteža elementā?
print("Is string 192 in the the first element of the tuple?", "192" in ip_port[0])

Is 80 in the tuple? True
Is 192 in the tuple? False
Is string 192 in the the first element of the tuple? True


In [6]:
# kortežu pakošana un atpakošana (tuple packing and unpacking)

# pakošana
t = 12345, 54321, 'hello!' # ar komatiem atdalītas vērtības tiek sapakotas kortežā
print(t) 

(12345, 54321, 'hello!')


In [8]:
# kortežu atpakošana
# kortežu atpakošana ir veids, 
# kā piešķirt korteža elementu vērtības vairākiem mainīgajiem

a, b, c = 1, 2, 3
print(a)
print(b)
print(c)

1
2
3


In [9]:
# ko darīt ja vēlamies atpakot tikai daļu korteža elementu?
# mēs varēsim izmantot * operatoru

numbers_tuple = tuple(range(10))
print(numbers_tuple)
first, second, *rest, tail = numbers_tuple
print(first, second, rest, tail) # piezīme: rest ir saraksts, nevis kortežs

# tātad šo kortežu atpakošanas paņemienu var izmantot arī sarakstiem

(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
0 1 [2, 3, 4, 5, 6, 7, 8] 9


In [10]:
# griešana darbojas arī uz kortežiem
# ja mēs nezinātu par atpakošanu, mēs varētu rīkoties šādi	
first = numbers_tuple[0]
second = numbers_tuple[1]
rest = numbers_tuple[2:-1] # sākot no 3. elementa līdz pēdējam elementam, to neieskaitot
tail = numbers_tuple[-1]
print(first, second, rest, tail) 
# piezīme: rest ir kortežs, nevis saraksts, tāpēc ka mēs griezām kortežu

0 1 (2, 3, 4, 5, 6, 7, 8) 9


In [12]:
# kortežu pakošana un atpakošana var būt noderīga, lai apmainītu divu mainīgo vērtības
a = 1985
b = 1955
print(a, b)
a, b = b, a # interpretators strāda no labās puses uz kreiso
# tehniski mēs veidojam kortežu labajā pusē, un tad atpakojam to kreisajā pusē
print(a, b)

1985 1955
1955 1985


In [13]:
# Korteži ir noteikta izmēra un šķietami nemainīgi
# tomēr, ja kortežā ir kāda maināma datu struktūra, kā piemēram saraksts, 
# jūs varat mainīt šīs datu struktūras saturu

# piemēram

# kortežs ar sarakstu kā vienu no elementiem
tup = (1,2,3,[4,5,6], 7, 8, 9)
print(tup)
# mēs varam pievienot elementu sarakstam, kas ir korteža elements
tup[3].append(700)
print(tup)
# mēs varam nomainīt elementu sarakstam, kas atrodas korteža iekšpusē
tup[3][0] = 400
print(tup)
# mēs varam sakārtot sarakstu, kas ir korteža elements
tup[3].sort()
print(tup)
# mēs arī varam izdzēst visus saraksta elementus
# pats saraksts paliek, bet tas ir tukšs
tup[3].clear()
print(tup)

# mēs nevaram nomainīt pašu elementu kortežā - tas ir nemainīgs
# piemēram, sarakstu nevar nomainīt ar citu sarakstu
try:
    tup[3] = [1,2,3] # te būs izņēmuma situācija
except TypeError as e:
    print(e)

(1, 2, 3, [4, 5, 6], 7, 8, 9)
(1, 2, 3, [4, 5, 6, 700], 7, 8, 9)
(1, 2, 3, [400, 5, 6, 700], 7, 8, 9)
(1, 2, 3, [5, 6, 400, 700], 7, 8, 9)
(1, 2, 3, [], 7, 8, 9)
'tuple' object does not support item assignment


## 3. tēma - funkcijas

### Funkcijas - programmu strukturēšanas pamata bloki

- Funkcijas var izmantot atkārtoti, tādējādi samazinot koda apjomu un padarot to lasāmāku un vieglāk uzturamu.

### Kas ir funkcija?

- Funkcija ir koda bloks, kas veic konkrētu uzdevumu, kad to izsauc.
- Funkcijai vajadzētu darīt vienu lietu un darīt to labi.
- Citiem vārdiem sakot, funkcijai vajadzētu būt vienam mērķim.
- Lielas funkcijas, kas veic daudz uzdevumu, ir grūti atkļūdot un uzturēt.

 
Funkcijai:

- [parasti] ir nosaukums (ar kuru to izsauc)
   - `fun_nosaukums()`
- var būt argumenti (vērtības, kuras funkcija saņem)
   - `fun_nosaukums(argumenti)`
- var būt atgriežamā vērtība(s)
   - `return vertiba`


In [15]:
# definēsim vienkāršu funkciju, kas izdrukā ziņojumu

def greet_user(): # funkcijas nosaukumam jābūt saprotamam un jāatspoguļo funkcijas būtība
    """Display a simple greeting."""
    # """ pēc funkcijas definīcijas ir docstring, kas apraksta, ko funkcija dara
    # tas nozīmē, ka ierakstot help(greet_user), jūs redzēsiet docstring
    # arī lielākā daļa IDE, kad jūs virs funkcijas nosaukuma pārvietosiet peli, jūs redzēsiet docstring
    # iespējams arī izveidot dokumentāciju ar Sphinx, kas izmanto docstring
    # sīkāk par docstring: https://www.python.org/dev/peps/pep-0257/
    print("Hello there!")

# neizsaucot funkciju, tā nedarbosies

In [16]:
# tagad mēs varēsim izsaukt funkciju
greet_user()

Hello there!


In [17]:
help(greet_user)

Help on function greet_user in module __main__:

greet_user()
    Display a simple greeting.



In [18]:
# Mēs varētu vēlēties izveidot pielāgotu sveiciena funkciju
# To var izdarīt, padodot tai vārdu un sveicienu

def greet(name, greeting): # name un greeting ir parametri
    print(f"{greeting} {name}")

# Tehniski, kad mēs definējam funkciju, mēs definējam parametrus
# Kad mēs izsaucam funkciju, mēs padodam argumentus
# Šie argumenti tiek piešķirti parametriem
greet("Uldis", "Hello from Valdis!") # Uldis and Hello from Valdis! ir argumenti

# bieži vien argumenti un parametri tiek izmantoti kā vārdi ar līdzīgu nozīmi

# mēs varam izsaukt funkciju arī ar citiem argumentiem
greet("Valdis", "Hi mom!")

Hello from Valdis! Uldis
Hi mom! Valdis


In [20]:
# Python mums ļauj izveidot funkcijas ar noklusētām vērtībām
# tas nozīmē, ka, ja mēs nepieškirsim vērtību šim parametram, tiks izmantota noklusētā vērtība
def default_greeting(text="World"):
    print("Hello", text)

default_greeting()
default_greeting("Python")
default_greeting(text="Latvia") # šeit mēs izmantojam vārdu, lai norādītu, kura vērtība tiek mainīta
# tas ir noderīgi, ja funkcijai ir vairāki parametri ar noklusētām vērtībām
# it īpaši svarīgi tas būtu, ja ir daudz bool tipa parametru.
# Savādāk būs grūti saprast, kuriem parametriem tiek piešķirta vērtība


Hello World
Hello Python
Hello Latvia


In [21]:
# Mūsu funkcijas var ne tikai saņemt argumentus, bet arī atgriezt vērtības.
# Tam izmanto return komandu.
# Ja mēs neizmantojam return komandu, funkcija atgriezīs None.

# Piemērs 1: atgriežam vienu vērtību
def add(a, b):
    return a + b

# Piemērs 2: atgriežam vairākas vērtības
def square_and_cube(x):
    return x ** 2, x ** 3 # mēs atgriežam kortežu ar diviem elementiem

def square_and_cube_list(x):
    return [x ** 2, x ** 3] # mēs atgriežam sarakstu ar diviem elementiem
    
print(add(2, 3))
print(square_and_cube(3))
print(square_and_cube_list(4))

5
(9, 27)
[16, 64]


In [22]:
# Jāpiemin ka Python interpretators nepārbauda funkciju argumentu datu tipus
print(add("Kaut kāds", " teksts"))

Kaut kāds teksts


In [23]:
# ja funkcija atgriež vērtību, mēs to varam saglabāt mainīgajā
add_result = add(1, 2)
print(add_result)
my_tuple = square_and_cube(3)
print(my_tuple)
square, cube = square_and_cube(3) # izmantosim atpakošanu
print(square, cube)


3
(9, 27)
9 27


### Brīdinājums: Nelietojiet maināmus objektus kā noklusējuma argumentus funkcijām

In [25]:
# Jāuzmantās definējot noklusētās vērtības
# 99.9% gadījumu jums vajadzētu izmantot nemainīgus objektus kā noklusētās vērtības
# savādak var sanākt neparedzēti rezultāti

def foo(bar=[]):
    bar.append("baz")
    return bar

print(foo())
print(foo()) 
# kas notika šeit?
# šeit tika izveidots tikai viens saraksts, un tas tika modificēts divreiz
# visticamāk, tas nav tas, ko jūs vēlaties
# tas ir tāpēc, ka noklusētās vērtības tiek piešķirtas funkcijas definīcijas laikā
print("*"*40)
# pamēģināsim vēlreiz
# šis būtu piemērotākais veids, kā to izdarīt, ja tiešam vajag sarakstu kā noklusēto vērtību
def foo_corrected(bar=None):
    if bar is None:
        bar = []
    bar.append("baz")
    return bar

print(foo_corrected())
print(foo_corrected())

# varējam arī padot jaunu sarakstu kā argumentu
print(foo([]))
print(foo([]))

# key takeway - use immutable objects as default values
# galvenais secinājums - izmantojiet nemainīgus objektus kā noklusētās vērtības, piemēram None, kortēžus, skaitļus

['baz']
['baz', 'baz']
****************************************
['baz']
['baz']
['baz']
['baz']


### Funkcijas ar patvaļīgu argumentu skaitu

In [27]:

# mēs varam izveidot funkciju, kas pieņem patvaļīgu argumentu skaitu
# šim nolūkam mēs varam izmantot *args
# args ir tikai mainīgā nosaukums, mēs varam izmantot jebkuru nosaukumu, ko vēlamies


def my_sum(*args):
    print(f"Supplied arguments: {args}")
    return sum(args)

def my_product(*args):
    print(f"Supplied arguments: {args}")
    product = 1 # svarīgi, jo reizināšanas vienība ir 1
    for arg in args:
        product *= arg
    return product

print(my_sum())
print(my_sum(1,2,3,4,5))
print("*"*40)
print(my_product())
print(my_product(1,2,3,4,5))



Supplied arguments: ()
0
Supplied arguments: (1, 2, 3, 4, 5)
15
****************************************
Supplied arguments: ()
1
Supplied arguments: (1, 2, 3, 4, 5)
120


In [28]:
# mēs varam kombinēt pozicionālos un mainīgos argumentus
def my_function(a, b, *args):
    """
    Šī funkcija ņem divus pozicionālos argumentus un mainīgu skaitu papildus argumentu
    un saskaita tos visus.
    Argumenti:
        a: pozicionāls arguments
        b: pozicionāls arguments
        args: kortežs ar mainīgu skaitu argumentu   
    Atgriež:
        Visu argumentu summu        
        """
    print(a, b, args)
    return a + b + sum(args)

my_function(1, 2, 3, 4, 5)


1 2 (3, 4, 5)


15

### Funkciju datu tipu norādīšana

In [30]:
# Mēs varam izmantot tipu norādījumus, lai norādītu funkcijas argumentu tipus
# Mēs varam izmantot tipu norādījumus, lai norādītu funkcijas atgriezto vērtību tipu
# Šie norādījumi ir neobligāti, bet tie var būt noderīgi dokumentācijas nolūkiem
# Python IGNORĒS šos tipu norādījumus, tie ir tikai dokumentācijai un ārējiem koda pārbaudes rīkiem - linteriem
# Kāpēc mēs izmantosim tipu norādījumus? kad mēs rakstām bibliotēku, ko izmantos citi cilvēki.
# Kad rakstām lielu programmu, ko mēs izmantosim ilgu laiku


def add_hints(a: int, b: int) -> int:
    return a + b

print(add_hints(1, 2))
print(add_hints('a', 'b')) # strāda, bet linter rīks brīdinās, ka datu tipi nesakrīt

# populārs linting rīks ir PyLance, kas nāk kopā ar Python paplašinājumu VS Code
# ir arī PyLint, Flake8, mypy utt.


3
ab


In [30]:
help(add_hints)

Help on function add_hints in module __main__:

add_hints(a: int, b: int) -> int



## 4. tēma - Vārdnīcas (dictionaries)

### Vārdnīcas - atslēgu-vērtību pāri

Galvenās vārdnīcu īpašības:
- Vārdnīcas ir nekārtotas
- Vārdnīcas ir maināmas(to vērtības var mainīt)	
- Vārdnīcas ir dinamiskas (to var paplašināt un sašaurināt)
- Vārdnīcu atslēgas ir unikālas un nemaināmas objekti (piemēram, teksts, skaitļi, nemaināmi korteži)
- Vārdnīcu vērtības var būt gandrīz jebkas (piemēram, virknes, skaitļi, saraksti, citas vārdnīcas, utt.)


In [1]:
# izveidosim tukšu vārdnīcu
my_dict = {}
print(my_dict)
# len() atgriež pāru skaitu vārdnīcā
print("Length of my_dict:", len(my_dict))

# izveidosim vārdnīcu ar diviem atslēgu-vērtību pāriem
my_dict = {'food': 'pizza', 'number': 42}
print(my_dict)
print("Length of my_dict:", len(my_dict))


{}
Length of my_dict: 0
{'food': 'pizza', 'number': 42}
Length of my_dict: 2


In [2]:
# mēs varam izmantot 'in' operatoru, lai pārbaudītu, vai mums ir kāda atslēga vārdnīcā
# pārbaudīsim, vai mums ir piens ledusskapī
print("Is there any milk in the fridge?", "milk" in my_dict) # O(1) laika sarežģītība
# kā būtu ar atslēgu 'food'?
print("Is there any food in the fridge?", "food" in my_dict)

Is there any milk in the fridge? False
Is there any food in the fridge? True


In [3]:
# pievienosim jaunu atslēgu-vērtību vārdnīcai
# pievienosim atslēgu pienu mūsu vārdnīcai

my_dict['milk'] = "Valmieras piens" # O(1) laika sarežģītība
print(my_dict)

{'food': 'pizza', 'number': 42, 'milk': 'Valmieras piens'}


In [5]:
# kā būtu ja mēs mainītu jau esošo atslēgu vērtību?
# nomainīsim vērtību atslēgai 'food'
my_dict['food'] = 'vegan salad'  # O(1) laika sarežģītība

In [36]:
# mēs varam apskatīt visas vērtības vārdnīcā
# šis objekts izskatās līdzīgi sarakstam, bet tas nav saraksts
my_dict.values()

dict_values(['vegan salad', 42, 'Valmieras piens'])

In [6]:
# mēs varam pārbaudīt vai kāda vērtība ir vārdnīcā
# kā būtu pārbaudīt, vai vārdnīcā ir vērtība 42?
# to mēs varēsim izdarīt, izmantojot 'in' operatoru kopā ar vārdnīcas values
# te būs O(n) laika sarežģītība, kur n ir atslēgu skaits vārdnīcā

print("Is there 42 in the dictionary values? ", 42 in my_dict.values()) # O(n) laika sarežģītība

Is there 42 in the dictionary values?  True


In [7]:
# kā būtu atrast vērtību, izmantojot atslēgu?	
# tas ir viens no galvenajiem iemesliem, kāpēc izmantot vārdnīcu
# atrast vērtību pēc atslēgas ir ātri - O(1) - konstanta laika sarežģītība - tas ir tas pats kas HashTable,HashMap,Map citās valodās
# atrast vērtību iterējot cauri visām vērtībām ir lēni - O(n) - lineāra laika sarežģītība

print("What food is in my fridge?", my_dict["food"])

What food is in my fridge? vegan salad


In [8]:
# kā būtu atrast vērtību pēc neeksistējošas atslēgas?
# tas izraisīs KeyError izņēmumu
try:
    print("My drink is", my_dict["drink"])
except KeyError:
    print("KeyError: drink is not in my_dict")

KeyError: drink is not in my_dict


In [9]:
# mēs būtu varējuši izmantot 'in' operatoru, lai pārbaudītu, vai atslēga ir vārdnīcā pirms mēģināt piekļūt vērtībai
if 'drink' in my_dict:
    print(my_dict['drink'])
else:
    print('drink is not in the dictionary')

# tomēr ir iespējams izmantot get() metodi, lai izdarītu to pašu
print(my_dict.get('drink')) # atgriezīs None, ja atslēga nav vārdnīcā
print(my_dict.get('drink', 'drink is not in the dictionary')) # otrais arguments ir noklusējuma vērtība, ja atslēga nav vārdnīcā
print(my_dict.get('food', 'food is not in the dictionary'))

drink is not in the dictionary
None
drink is not in the dictionary
vegan salad


In [12]:
# mēs varam apskatīt visas atslēgu-vērtību pārus vārdnīcā kā sarakstam līdzīgu objektu, kurš satur kortēžus
print(my_dict.items()) 
my_dict_items = list(my_dict.items()) # varam to pārveidot par sarakstu ar list() funkciju
print(my_dict_items)

dict_items([('food', 'vegan salad'), ('number', 42), ('milk', 'Valmieras piens')])
[('food', 'vegan salad'), ('number', 42), ('milk', 'Valmieras piens')]


In [13]:
# iterēšana cauri vārdnīcai
for key in my_dict:
    print(key, my_dict[key])

print()

# alternatīva ir izmantot items() metodi, kas atgriež atslēgu-vērtību pārus
for key, value in my_dict.items(): # key, value are arbitrary names, could be k,v etc
    print(key, value)

food vegan salad
number 42
milk Valmieras piens

food vegan salad
number 42
milk Valmieras piens


In [14]:
# jāuzmanās, lai nemainītu vārdnīcas izmēru, kamēr iterējam cauri tai
# labāk izmantot kopiju vārdnīcas, ja plānojam mainīt oriģinālo vārdnīcu	
for key,value  in my_dict.copy().items():
    if value == 42:
        del my_dict[key] # izdzēsīsim atslēgu, ja vērtība ir 42

print(my_dict)

{'food': 'vegan salad', 'milk': 'Valmieras piens'}


In [15]:
my_dict2 = {}

# šāda pieeja nedarbosies, jo saraksts ir ir mainīgs objekts
key2 = [1, 2, 3]
value2 = "teksts"
try:
    my_dict2[key2] = value2
except TypeError as e:
    print(e)

unhashable type: 'list'


In [16]:
# savukārt kortežs kurā ir skaitļi ir nemainīgs objekts un tas darbosies kā atslēga
key2 = (1, 2, 3)
my_dict2[key2] = value2

print(my_dict2)

{(1, 2, 3): 'teksts'}


In [17]:
my_dict2[(2024,9,16)] = "today"
print(my_dict2)

{(1, 2, 3): 'teksts', (2024, 9, 16): 'today'}


### Vārdnīcu metodes

Šeit ir metodes, ko var izmantot ar vārdnīcām
- .keys() atgriež sarakstu ar visām atslēgām vārdnīcā
- .values() atgriež sarakstu ar visām vērtībām vārdnīcā
- .items() atgriež sarakstu ar pāriem atslēga-vērtība vārdnīcā
- .get(key) atgriež vērtību, kas saistīta ar atslēgu, citādi None
- .get(key, default) atgriež vērtību, kas saistīta ar atslēgu, citādi default
- .update(other_dictionary) pievieno visus atslēgu-vērtību pārus no other_dictionary vārdnīcai
- .pop(key) noņem atslēgu-vērtību pāri, kas saistīts ar atslēgu, no vārdnīcas
- .popitem() noņem nejauši izvēlētu atslēgu-vērtību pāri no vārdnīcas
- .clear() noņem visus atslēgu-vērtību pārus no vārdnīcas

Pilns metožu saraksts: https://docs.python.org/3/library/stdtypes.html#dict

Atslēgu secība vārdnīcās ir tāda pati, kādā tās tika pievienotas(kopš Python 3.6):

Ja jums vajag sakārtotu vārdnīcu, jums vajadzētu izmantot OrderedDict no collections moduļa.
Tikai jāņem vērā ka OrderedDict būs lēnāka par parastu vārdnīcu. 
Daudzas operācijas būs O(log n) laika, nevis O(1) laikā kā parastās vārdnīcas.

Šeit vietā analoģija ar C++ Map kura ir līdzīga Python sakārtotai vārdnīcai, un C++ unordered_map kura ir nekārtota vārdnīca.

Parasti mums nav nepieciešams zināt atslēgu secību, bet ir svarīgi zināt, ka tā ir.

### Topic 4: - Sets

#### Sets in Python are unordered collections of unique elements

Key features of sets:
- sets are unordered
- elements in sets are unique
- sets are mutable
- sets can be used to perform mathematical set operations like union, intersection, symmetric difference etc.

Primary uses of sets:
- removing duplicate entries
- fast membership testing
- mathematical set operations

In [52]:
# let's create a set
set_1 = {1, 2, 3, 4, 5,2,1,4} # note we use same syntax as for dictionaries but without key-value pairs
print(set_1) # all duplicates are removed, order might look different than what we entered, it is not guaranteed

{1, 2, 3, 4, 5}


In [53]:
# we can check in constant time if an element is in a set
print(1 in set_1) # this is O(1) time

True


In [54]:
# let's update our set
set_1.update(range(2,12))
print(set_1) # again for small numbers sets might look ordered but they are not
set_1.update(range(-5,5)) # might not look so ordered anymore
print(set_1)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, -1, -5, -4, -3, -2}


In [55]:
# so sets can be used to remove duplicates from a list
my_list = [1, 2, 3, 4, 5, 5, 5, 6, 7, 7]
my_set = set(my_list)
print(my_set)
# go back to list
my_list = list(my_set)
print(my_list)

{1, 2, 3, 4, 5, 6, 7}
[1, 2, 3, 4, 5, 6, 7]


In [56]:
# finally sets can be used for various set operations like union, intersection, difference, etc.

n_3_7 = set(range(3, 8))
n_5_9 = set(range(5, 10))
print(n_3_7)
print(n_5_9)
# union 
# union means all elements from both sets without duplicates
print(n_3_7.union(n_5_9))
# alternative union
print(n_3_7 | n_5_9) # syntactic sugar

{3, 4, 5, 6, 7}
{5, 6, 7, 8, 9}
{3, 4, 5, 6, 7, 8, 9}
{3, 4, 5, 6, 7, 8, 9}


In [ ]:
# intersection of two sets
# intersection means common elements

print(n_3_7.intersection(n_5_9))
# alternative syntax
print(n_3_7 & n_5_9)


In [57]:
# we also have difference operation
# difference() - returns a set that contains the difference between two sets
# this means return all items that are in the first set, and not in the second set

print(n_3_7.difference(n_5_9))
# difference is not commutative
print(n_5_9.difference(n_3_7))
# there is also shorthand for difference
print(n_3_7 - n_5_9)

# there is also symmetric_difference() - returns a set that contains all items from both sets, except items that are present in both sets
print(n_3_7.symmetric_difference(n_5_9))
# shorthand
print(n_3_7 ^ n_5_9)

{3, 4}
{8, 9}
{3, 4}
{3, 4, 8, 9}
{3, 4, 8, 9}


---

There are more set methods such as isdisjoint, issubset, issuperset, union, intersection, difference, symmetric_difference, ...

Additional information: https://docs.python.org/3/library/stdtypes.html#set

---

## Lesson Overview

What have we learned?

* Tuples - also known as immutable lists
* Functions - the basic building blocks of programs
* Dictionaries - key-value pairs
* Sets - unordered collections of unique elements with support for set operations

#### Topic 1 - tuple mini exercise

In [ ]:
# create a tuple representing you as a person and your age
# from the tuple create a new tuple with your age increased by 1 and your last and first name swapped

## Exercises for further practice

### Mini Exercises

---

Write a function that takes in a dictionary and value and returns a list of tuples of all keys and values in the dictionary whose value matches the value passed in.
- example:
   - input: `{'a': 1, 'b': 2, 'c': 3, 'd': 2}, 2`
   - output: `[('b', 2), ('d', 2)]`

Note: this can be done very efficiently using list comprehension.

--- 

Write a function called `sum_squares` that takes a list of numbers and returns the sum of the squares of all the numbers.

---

### Exercise 1

Write a function that takes a string and counts the number of times each character appears in the string. Return a list of tuples, where each tuple contains a character and the number of times it appears in the string.

Bonus if you can supply top n most frequent characters to return.

### Exercise 2 - is_pangram

Write a function called `is_pangram` that will return `True` if a given string is a pangram, otherwise it will return `False`. A pangram is a sentence that contains every letter of the alphabet at least once. For example, the sentence `"The quick brown fox jumps over the lazy dog"` is a pangram because it uses the letters `a-z` at least once (case is irrelevant).

Bonus if you can ignore punctuation and whitespace.

Extra bonus if you can supply custom alphabets to check against. This would allow you to check if a sentence is a pangram in a language other than English.

## Additional Resources

### Topic 1 - tuples

- [Tuple official docs](https://docs.python.org/3/library/stdtypes.html#tuples)

### Topic 2 - functions

- [Functions built ins official docs](https://docs.python.org/3/library/functions.html)
- [Automate the boring stuff with Python: Functions](https://automatetheboringstuff.com/2e/chapter3/)

### Topic 3 - dictionaries

- [Dictionaries official docs](https://docs.python.org/3/library/stdtypes.html#dict)
- [Automate the boring stuff with Python: Dictionaries](https://automatetheboringstuff.com/2e/chapter5/)

### Topic 4 - sets

- [Sets official docs](https://docs.python.org/3/library/stdtypes.html#set)


## Additional Information

- [OpenPyXL library](https://openpyxl.readthedocs.io/en/stable/tutorial.html) for working with Excel files

- Python 3.12 was released on October 2, 2023
  - [Python 3.12: New features](https://realpython.com/python312-new-features/)
  - [What's new in Python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html)
 
